content/blog/2020-12-18-db-compiler/index.md

```diff
@@ -0,0 +1,293 @@
++++
+title = "Compilation-based Execution Engine for a Database"
+[extra]
+bio = """
+ Ankush Rayabhari is a first-year MS student in Computer Science at Cornell who
+ is interested in databases and compilers.
+"""
+[[extra.authors]]
+name = "Ankush Rayabhari"
++++
+
+My project was to implement a compilation-based execution engine for a database. The code is located [here](https://github.com/ankushrayabhari/kushdb/).
+
+## Introduction
+
+Existing database systems initially convert a SQL query to a tree of relational algebra operators and then interpret that plan. In the past, this wasn't a bottleneck as the main cost of processing data was reading and writing data to disk. Since the amount of CPU time spent processing data was negligible compared to the disk access time, the overheads of interpretation didn't play a large role.
+
+Nowadays, databases have access to faster storage and large amounts of main memory which means that the CPU time spent processing data is much more valuable. To reduce the overhead of interpretation, we can directly compile trees of operators down to binary and then execute them. This projects explores implementing such a compiler.
```

Comment:
Might be worth linking to/citing some sort of canonical reference on query compilation? Like, is [this one](https://15721.courses.cs.cmu.edu/spring2018/papers/03-compilation/shaikhha-sigmod2016.pdf) any good?

Reply:
I've linked some slides there.